src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -3306,6 +3306,13 @@ private uint getClassDomainID(CORINFO_CLASS_STRUCT_* cls, ref void* ppIndirectio
             }
         }
 
+#pragma warning disable CA1822 // Mark members as static
+        private void* getFrozenHandleFromInitedStaticField(CORINFO_FIELD_STRUCT_* field)
```

Comment:
@MichalStrehovsky Could you please suggest how to implement this for pre-inited fields?

Reply:
@jkotas is it supposed to work in AOT scenarios where all types aren't initialized at the compilation time? Perhaps with some interprocedure analysis I assume that can prove that no reflection is used to change its value? In current Main it's not currently handled it seems:

```csharp
public class Program
{
    static void Main() => Console.WriteLine(GetFoo());

    static readonly int s_foo = 42;

    [MethodImpl(MethodImplOptions.NoInlining)]
    static int GetFoo() => s_foo;
}
```
ILC (Main) produces:
```asm
; Method Program:GetFoo():int
G_M22732_IG01:              
       4883EC28             sub      rsp, 40
G_M22732_IG02:              
       E800000000           call     CORINFO_HELP_READYTORUN_STATIC_BASE
       8B00                 mov      eax, dword ptr [rax]
G_M22732_IG03:              
       4883C428             add      rsp, 40
       C3                   ret      
; Total bytes of code: 16
```

Reply:
The `CORINFO_HELP_READYTORUN_STATIC_BASE` helper in your example just returns a constant address using `lea rax, [rip + 0xXXXX]`. Getting this inlined is separate optimization - it is the second item from https://github.com/dotnet/runtime/issues/64242.

The storage for `s_foo` is allocated on the frozen heap that is part of the binary in your example. We know that it is always going to have value 42 during aot compilation and `getReadonlyStaticFieldValue` should tell it to the JIT.

Second example to consider:
```
using System.Runtime.CompilerServices;

public class Program
{
    static void Main()
    {
        // `s.GetType() == typeof(Action)` is optimized into `true` in native AOT, just like it is optimized with tiered JIT 
        for (;;) Console.WriteLine(s.GetType() == typeof(Action));
    }

    // This is interpreted at aot compile time and the `Action` instance is allocated on frozen heap that is part of the aot binary
    // It requires getObjectType helper handling of non-string frozen objects
    static readonly object s = new Action(Main);
}
```

Reply:
Ah good to know, we should add an SPMI collection for NativeAOT to track improvements

Reply:
@jkotas @MichalStrehovsky so I put an initial impl that works as expected in jit if I just return 42 from ILC - https://github.com/dotnet/runtime/pull/76112/commits/1c584a19e1b63a17d88db22e5f714931de719e04
but I don't understand how to get the actual field value in this case

Reply:
It should be something like this:
```
var preinitManager = _compilation.NodeFactory.PreinitializationManager;
if (preinitManager.IsPreinitialized(type))
{
    var preinitializationInfo = preinitManager.GetPreinitializationInfo(_type);
    var value = preinitializationInfo.GetFieldValue(field);
    ... marshal the value into what the JIT/EE interface returns ... 
}
```
Michal is the best person to comment on the details.

Reply:
Thanks, that's exactly what I needed, pushed a change to handle all primitive fields, new codegen for `GetFoo` with NativeAOT:
```asm
; Assembly listing for method Program:GetFoo():int
G_M22732_IG01:              
G_M22732_IG02:              
       B82A000000           mov      eax, 42
G_M22732_IG03:              
       C3                   ret
```

Reply:
And pushed a change for string literals

Reply:
```csharp
static readonly object f = null;

[MethodImpl(MethodImplOptions.NoInlining)]
static object GetFoo() => f;
```
I assume Roslyn removes `= null` so this no longer passes `IsPreinitialized` check and is not optimized to null. But presumably it's a rare case.

Reply:
Yeah, this doesn't work because there's no cctor. We could fix this, but also it would be extra code and this doesn't look common enough (you really have to be in a situation where there's a readonly field and no cctor). I would maybe even expect Roslyn to generate a warning on it.